models/percona_sso_model_helpers.go

```diff
@@ -17,32 +17,128 @@
 package models
 
 import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
 	"github.com/pkg/errors"
 	"gopkg.in/reform.v1"
 )
 
+var perconaSSOMtx sync.Mutex
+
 // GetPerconaSSODetails returns PerconaSSODetails if there are any, error otherwise.
-func GetPerconaSSODetails(q *reform.Querier) (*PerconaSSODetails, error) {
-	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsView, "")
+// Access token is automatically refreshed if it is expired.
+// Get, check eventually refresh done in one tx.
+func GetPerconaSSODetails(ctx context.Context, q *reform.Querier) (*PerconaSSODetails, error) {
+	perconaSSOMtx.Lock()
+	defer perconaSSOMtx.Unlock()
+
+	ssoDetails, err := q.SelectOneFrom(PerconaSSODetailsTable, "")
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to get Percona SSO Details")
 	}
-	return ssoDetails.(*PerconaSSODetails), nil
+
+	details := ssoDetails.(*PerconaSSODetails)
+	if details.isAccessTokenExpired() {
+		refreshedToken, err := details.refreshAndGetAccessToken(ctx, q)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to insert Percona SSO Details")
+		}
+		details.AccessToken = refreshedToken
+	}
+
+	return details, nil
+}
+
+// GetPerconaSSOAccessToken returns PerconaSSOAccessToken if there is GetPerconaSSODetails, error otherwise.
+func GetPerconaSSOAccessToken(ctx context.Context, q *reform.Querier) (*PerconaSSOAccessToken, error) {
```

Comment:
why do we have a separate getter method if `ssoDetails.AccessToken ` is an exported field?

Reply:
I would suggest to return string only, so this method has a purpose. Without it, we'd have to write `ssoDetails.AccessToken.AccessToken` to get the string token we actually need. What do you guys think?

Reply:
but even just the string `AccessToken` is an exported field, if we want a method to get this value then we should make these fields unexported don;t you think?

Reply:
Well then we can't use reform. Because it can not insert into unexported field.

Reply:
Or use some proxy struct to fetch it.

Reply:
Removed whole separated method to get access token.